Bug: [no-floating-promises] TypeOrValueSpecifier (allowForKnownSafePromises) doesn't match FastifyRegister

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Following https://github.com/fastify/fastify/pull/5506 -> https://fastify.dev/docs/v5.1.x/Reference/TypeScript#plugins, the [`@typescript-eslint/no-floating-promises` `allowForKnownSafePromises` option](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises) should work to allowlist Fastify promises when the [`checkThenables` option](https://typescript-eslint.io/rules/no-floating-promises/#checkthenables) would cause them to report.

However, in https://github.com/typescript-eslint/examples/pull/22, I'm seeing that `FastifyRegister` is still being reported.

```ts
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [
        { "from": "package", "name": "FastifyInstance", "package": "fastify" },
        { "from": "package", "name": "FastifyRegister", "package": "fastify" },
        { "from": "package", "name": "SafePromiseLike", "package": "fastify" },
      ]
    }]
  }
}
```

```ts
// imported from "fastify" and "fastify-cookie"
fastify.register(FastifyCookie);
```

```plaintext
  11:1  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

I debugged into this a bit and found that:
* In `specifierNameMatches`, the `FastifyRegister` name is correctly being matched ✅ 
* However, in `typeMatchesSpecifier`, the `symbol` is `undefined`

I haven't investigated deeply enough yet to understand why the type wouldn't have a symbol. I don't think this is a user misconfiguration case. The repro is pretty straightforward; its ESlint config uses `projectServices` and its TSConfig has `NodeNext` for both `module` and `moduleResolution`.

Note that Fastify types use `PromiseLike` rather than `Promise`, so the [`checkThenables` option](https://typescript-eslint.io/rules/no-floating-promises/#checkthenables) must be enabled to show the lint complaints.

### Reproduction Repository Link

https://github.com/typescript-eslint/examples/pull/22

### Repro Steps

```shell
git clone https://github.com/typescript-eslint/examples
cd examples
gh pr checkout 22
npm i
cd packages/fastify-floating-promises
npm run lint
```

### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.18.0` |
| `TypeScript`                           | `5.7.2` |
| `ESLint`                               | `9.16.0` |

💖

--- a/packages/type-utils/src/TypeOrValueSpecifier.ts
+++ b/packages/type-utils/src/TypeOrValueSpecifier.ts
@@ -167,42 +167,50 @@ export function typeMatchesSpecifier(
   specifier: TypeOrValueSpecifier,
   program: ts.Program,
 ): boolean {
-  if (tsutils.isIntrinsicErrorType(type)) {
-    return false;
-  }
-  if (tsutils.isIntersectionType(type)) {
-    const somePartMatches = tsutils
-      .intersectionTypeParts(type)
-      .some(part => typeMatchesSpecifier(part, specifier, program));
-
-    if (somePartMatches) {
-      return true;
+  const wholeTypeMatches = ((): boolean => {
+    if (tsutils.isIntrinsicErrorType(type)) {
+      return false;
     }
+    if (typeof specifier === 'string') {
+      return specifierNameMatches(type, specifier);
+    }
+    if (!specifierNameMatches(type, specifier.name)) {
+      return false;
+    }
+    const symbol = type.getSymbol() ?? type.aliasSymbol;
+    const declarations = symbol?.getDeclarations() ?? [];
+    const declarationFiles = declarations.map(declaration =>
+      declaration.getSourceFile(),
+    );
+    switch (specifier.from) {
+      case 'file':
+        return typeDeclaredInFile(specifier.path, declarationFiles, program);
+      case 'lib':
+        return typeDeclaredInLib(declarationFiles, program);
+      case 'package':
+        return typeDeclaredInPackageDeclarationFile(
+          specifier.package,
+          declarations,
+          declarationFiles,
+          program,
+        );
+    }
+  })();
+
+  if (wholeTypeMatches) {
+    return true;
   }
-  if (typeof specifier === 'string') {
-    return specifierNameMatches(type, specifier);
+
+  if (
+    tsutils.isIntersectionType(type) &&
+    tsutils
+      .intersectionTypeParts(type)
+      .some(part => typeMatchesSpecifier(part, specifier, program))
+  ) {
+    return true;
   }
-  if (!specifierNameMatches(type, specifier.name)) {
-    return false;
-  }
-  const symbol = type.getSymbol() ?? type.aliasSymbol;
-  const declarations = symbol?.getDeclarations() ?? [];
-  const declarationFiles = declarations.map(declaration =>
-    declaration.getSourceFile(),
-  );
-  switch (specifier.from) {
-    case 'file':
-      return typeDeclaredInFile(specifier.path, declarationFiles, program);
-    case 'lib':
-      return typeDeclaredInLib(declarationFiles, program);
-    case 'package':
-      return typeDeclaredInPackageDeclarationFile(
-        specifier.package,
-        declarations,
-        declarationFiles,
-        program,
-      );
-  }
+
+  return false;
 }
 
 export const typeMatchesSomeSpecifier = (

--- a/packages/type-utils/src/TypeOrValueSpecifier.ts
+++ b/packages/type-utils/src/TypeOrValueSpecifier.ts
@@ -170,6 +170,15 @@ export function typeMatchesSpecifier(
   if (tsutils.isIntrinsicErrorType(type)) {
     return false;
   }
+  if (tsutils.isIntersectionType(type)) {
+    const somePartMatches = tsutils
+      .intersectionTypeParts(type)
+      .some(part => typeMatchesSpecifier(part, specifier, program));
+
+    if (somePartMatches) {
+      return true;
+    }
+  }
   if (typeof specifier === 'string') {
     return specifierNameMatches(type, specifier);
   }
@@ -200,18 +209,5 @@ export const typeMatchesSomeSpecifier = (
   type: ts.Type,
   specifiers: TypeOrValueSpecifier[] = [],
   program: ts.Program,
-): boolean => {
-  if (
-    specifiers.some(specifier => typeMatchesSpecifier(type, specifier, program))
-  ) {
-    return true;
-  }
-
-  if (tsutils.isIntersectionType(type)) {
-    return tsutils
-      .intersectionTypeParts(type)
-      .some(part => typeMatchesSomeSpecifier(part, specifiers, program));
-  }
-
-  return false;
-};
+): boolean =>
+  specifiers.some(specifier => typeMatchesSpecifier(type, specifier, program));

--- a/packages/type-utils/src/TypeOrValueSpecifier.ts
+++ b/packages/type-utils/src/TypeOrValueSpecifier.ts
@@ -200,5 +200,18 @@ export const typeMatchesSomeSpecifier = (
   type: ts.Type,
   specifiers: TypeOrValueSpecifier[] = [],
   program: ts.Program,
-): boolean =>
-  specifiers.some(specifier => typeMatchesSpecifier(type, specifier, program));
+): boolean => {
+  if (
+    specifiers.some(specifier => typeMatchesSpecifier(type, specifier, program))
+  ) {
+    return true;
+  }
+
+  if (tsutils.isIntersectionType(type)) {
+    return tsutils
+      .intersectionTypeParts(type)
+      .some(part => typeMatchesSomeSpecifier(part, specifiers, program));
+  }
+
+  return false;
+};

--- a/packages/type-utils/src/typeOrValueSpecifiers/specifierNameMatches.ts
+++ b/packages/type-utils/src/typeOrValueSpecifiers/specifierNameMatches.ts
@@ -1,7 +1,5 @@
 import type * as ts from 'typescript';
 
-import * as tsutils from 'ts-api-utils';
-
 export function specifierNameMatches(
   type: ts.Type,
   names: string | string[],
@@ -17,10 +15,6 @@ export function specifierNameMatches(
 
   if (names.some(item => candidateNames.includes(item))) {
     return true;
-  }
-
-  if (tsutils.isIntersectionType(type)) {
-    return type.types.some(subType => specifierNameMatches(subType, names));
   }
 
   return false;

--- a/packages/type-utils/tests/TypeOrValueSpecifier.test.ts
+++ b/packages/type-utils/tests/TypeOrValueSpecifier.test.ts
@@ -6,7 +6,7 @@ import path from 'node:path';
 
 import type { TypeOrValueSpecifier } from '../src/TypeOrValueSpecifier';
 
-import { typeMatchesSomeSpecifier, typeOrValueSpecifiersSchema } from '../src';
+import { typeMatchesSpecifier, typeOrValueSpecifiersSchema } from '../src';
 
 describe('TypeOrValueSpecifier', () => {
   describe('Schema', () => {
@@ -122,7 +122,7 @@ describe('TypeOrValueSpecifier', () => {
     ])("doesn't match a malformed package specifier: %s", runTestNegative);
   });
 
-  describe('typeMatchesSomeSpecifier', () => {
+  describe('typeMatchesSpecifier', () => {
     function runTests(
       code: string,
       specifier: TypeOrValueSpecifier,
@@ -143,9 +143,9 @@ describe('TypeOrValueSpecifier', () => {
               .id,
           ),
         );
-      expect(
-        typeMatchesSomeSpecifier(type, [specifier], services.program!),
-      ).toBe(expected);
+      expect(typeMatchesSpecifier(type, specifier, services.program!)).toBe(
+        expected,
+      );
     }
 
     function runTestPositive(

--- a/packages/type-utils/tests/TypeOrValueSpecifier.test.ts
+++ b/packages/type-utils/tests/TypeOrValueSpecifier.test.ts
@@ -6,7 +6,7 @@ import path from 'node:path';
 
 import type { TypeOrValueSpecifier } from '../src/TypeOrValueSpecifier';
 
-import { typeMatchesSpecifier, typeOrValueSpecifiersSchema } from '../src';
+import { typeMatchesSomeSpecifier, typeOrValueSpecifiersSchema } from '../src';
 
 describe('TypeOrValueSpecifier', () => {
   describe('Schema', () => {
@@ -122,7 +122,7 @@ describe('TypeOrValueSpecifier', () => {
     ])("doesn't match a malformed package specifier: %s", runTestNegative);
   });
 
-  describe('typeMatchesSpecifier', () => {
+  describe('typeMatchesSomeSpecifier', () => {
     function runTests(
       code: string,
       specifier: TypeOrValueSpecifier,
@@ -143,9 +143,9 @@ describe('TypeOrValueSpecifier', () => {
               .id,
           ),
         );
-      expect(typeMatchesSpecifier(type, specifier, services.program!)).toBe(
-        expected,
-      );
+      expect(
+        typeMatchesSomeSpecifier(type, [specifier], services.program!),
+      ).toBe(expected);
     }
 
     function runTestPositive(
@@ -395,13 +395,64 @@ describe('TypeOrValueSpecifier', () => {
       [
         `
           type SafePromise = Promise<number> & { __safeBrand: string };
-          type JoinedPromise = SafePromise & {};
+          type ResultType = { foo: 'bar' };
+          type Test = SafePromise & ResultType;
         `,
-        { from: 'file', name: ['JoinedPromise'] },
+        { from: 'file', name: ['ResultType'] },
       ],
     ])(
       'matches a matching type specifier for an intersection type: %s',
       runTestPositive,
+    );
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      [
+        `
+          declare module "node:test" {
+            type SafePromise = Promise<undefined> & { __safeBrand: string };
+            type ItResult = { foo: 'bar' };
+
+            export function it(): SafePromise & ItResult;
+          }
+
+          import { it } from "node:test";
+
+          type Test = ReturnType<typeof it>;
+        `,
+        {
+          from: 'package',
+          name: ['ItResult'],
+          package: 'node:test',
+        },
+      ],
+    ])(
+      'matches a matching package specifier for an intersection type: %s',
+      runTestPositive,
+    );
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      [
+        `
+          declare module "node:test" {
+            type SafePromise = Promise<undefined> & { __safeBrand: string };
+            type ItResult = { foo: 'bar' };
+
+            export function it(): SafePromise & ItResult;
+          }
+
+          import { it } from "node:test";
+
+          type Test = ReturnType<typeof it>;
+        `,
+        {
+          from: 'package',
+          name: ['Result'],
+          package: 'node:test',
+        },
+      ],
+    ])(
+      "doesn't match a mismatched package specifier for an intersection type: %s",
+      runTestNegative,
     );
 
     it("does not match a `declare global` with the 'global' package name", () => {

